### DISCUSSION TITLE
Did the UUID column behavior change in 2.0?

### DISCUSSION THREAD

**Discussion by mkismy** [sqlalchemy/sqlalchemy#10336] [Usage Questions]
I am currently upgrading my project to 2.0 and found some weird behavior.

When inserting uuid to a UUID column, the result differ on sqlalchemy 1.x and 2.x

On 1.x, the result will be like `aaaaaaaa-bbbb-cccc-dddd-eeeeeeee` with hyphens. On 2.x, it would be like `aaaaaaaabbbbccccddddeeeeeeee` without hyphens.

However, I found that when inserting the same uuid to a String column, both uuid in 1.x and 2.x had hyphens.

![image](https://github.com/sqlalchemy/sqlalchemy/assets/8445575/15caf248-f3e5-4b1f-bd0f-758a9e387b4f)

I couldn't find about this behavior change in the docs but is this a expected behavior?
And if so, can someone help me for a workaround to retain inserting the 1.x-style uuid with hyphens to a UUID column

Tested with python 3.11 + sqlalchemy 1.4.49 & 2.0.20 + Mariadb 10.5.19

```sql
CREATE TABLE `events` (
  `id` bigint(20) unsigned NOT NULL AUTO_INCREMENT,
  `uuid` char(36) NOT NULL,
  `uuid_string` varchar(191) NOT NULL,
) ENGINE=InnoDB AUTO_INCREMENT=129 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;

```

```python
import uuid

from sqlalchemy import BigInteger, Column, String, create_engine
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import declarative_base, scoped_session, sessionmaker

Base = declarative_base()

class Event(Base):
    __tablename__ = "events"

    id = Column(BigInteger, primary_key=True, index=True)
    uuid = Column(UUID(as_uuid=True), default=uuid.uuid4)
    uuid_string = Column(String(191))

engine = create_engine("mysql://user:pass@address/database")
session = scoped_session(sessionmaker(engine))

uuid = uuid.uuid4()

event = Event(
    uuid=uuid,
    uuid_string=uuid
)

session.add(event)
session.commit()
```

**Comment by CaselIT** [comment]
Hi,

You are using mysql with a type that's specific to postgresql. Don't do that :). In general this will not work, and the fact that it did in 1.x was coincidental.

You can use the generic https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.Uuid

**Reply by mkismy** [reply]
Hi, 

Yes, I understand that it's not appropriate and trying to update that too.
However using the newly introduced `Uuid` has no change on the outcome. It inserts uuids without hyphens.

Are there any workaround to keep using uuid with hyphens?

**Reply by CaselIT** [reply]
seems not at the moment. 
the compile was hardcoded to strip the hyphen here https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3871/2..3. @zzzeek do you remember why it was changed like this? It seems that it also a changes how the 1.4 pg type was saved when `native_uuid=False`.

> Are there any workaround to keep using uuid with hyphens?

You can use https://docs.sqlalchemy.org/en/20/core/custom_types.html#sqlalchemy.types.TypeDecorator to customize the UUID type

**Comment by zzzeek** [comment]
> seems not at the moment. the compile was hardcoded to strip the hyphen here https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3871/2..3. @zzzeek do you remember why it was changed like this? 

that's only the second revision of a single patch, this behavior was never "changed".  the behavior as in our original example at https://docs.sqlalchemy.org/en/20/core/custom_types.html#backend-agnostic-guid-type is to fit the string into 32 characters, so the hyphens have to be stripped.


> It seems that it also a changes how the 1.4 pg type was saved when `native_uuid=False`.

the pg type uses UUID in all cases, never CHAR(32), so it's always "saved" as a UUID, the presence of hyphens on a type going into the database does not matter to the drivers here.

The result behavior is that of the driver, it has not changed for the PostgreSQL UUID type from 1.4 to 2.0:

```py
import uuid

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import MetaData
from sqlalchemy import select
from sqlalchemy import Table
from sqlalchemy.dialects.postgresql import UUID


t = Table('t', MetaData(), Column('data', UUID(as_uuid=False)))

e = create_engine("postgresql://scott:tiger@localhost/test")
t.drop(e, checkfirst=True)
t.create(e, checkfirst=True)

with e.begin() as conn:
    conn.execute(
        t.insert(), {"data": str(uuid.uuid4())}
    )

    print(conn.scalar(select(t.c.data)))
```

1.4:

```
5a2ed9ab-7ba5-48ed-b0b5-5cc2169ca2ae
```

2.0:

```
64a2b11e-e418-4db3-b628-5b33de2bddda
```


This is the same behavior for the base `sqlalchemy.UUID` object  including on MariaDB:

```py
import uuid

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import MetaData
from sqlalchemy import select
from sqlalchemy import Table
from sqlalchemy import UUID


t = Table("t", MetaData(), Column("data", UUID(as_uuid=False)))

for e in (
    create_engine("postgresql://scott:tiger@localhost/test"),
    create_engine("mysql://scott:tiger@mariadb108/test"),
):
    t.drop(e, checkfirst=True)
    t.create(e, checkfirst=True)

    with e.begin() as conn:
        conn.execute(t.insert(), {"data": str(uuid.uuid4())})

        print(conn.scalar(select(t.c.data)))

```

```
ad66d072-246a-4e3c-b986-f1d610c527e2
5f06b72a-220e-40e7-b25c-1f5372b41481

```


so nothing has changed anywhere here, except as noted the undefined behavior of using the PostgreSQL UUID datatype on MariaDB.

Using dashes with a char(36) persistence format is an unusual use case so the recipe at https://docs.sqlalchemy.org/en/20/core/custom_types.html#backend-agnostic-guid-type should be adapted for CHAR(36) which will work predictably under all circumstances.

**Reply by CaselIT** [reply]
> the pg type uses UUID in all cases, never CHAR(32), so it's always "saved" as a UUID, the presence of hyphens on a type going into the database does not matter to the drivers here.

indeed on 1.4 there is no `native_uuid=False` on the pg type.

> Using dashes with a char(36) persistence format is an unusual use case so the recipe at https://docs.sqlalchemy.org/en/20/core/custom_types.html#backend-agnostic-guid-type should be adapted for CHAR(36) which will work predictably under all circumstances.

yes, it should note that v2 includes a uuid type natively, and maybe present this recipe to use dashes instead of the collapsed hex

**Reply by CaselIT** [reply]
Interesting we seem to not render UUID on mariadb https://mariadb.com/kb/en/uuid-data-type/

probably a 2.1 thing

**Reply by CaselIT** [reply]
created https://github.com/sqlalchemy/sqlalchemy/issues/10339 for it

**Comment by CaselIT** [comment]
@mkismy I've pushed an updated example here that does what you want. https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4864

Once merged you should also see it in the documentation

**Reply by mkismy** [reply]
Thank you! I'll try the custom type 🙌
